src/analysis/cache.Rmd

```diff
@@ -0,0 +1,313 @@
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$LLC_load_misses, t$opcode)), type="l", xlab="op_count", ylab="LLC_load_misses",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$L1_icache_loads, t$opcode)), type="l", xlab="op_count", ylab="L1_icache_loads",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$L1_icache_load_misses, t$opcode)), type="l", xlab="op_count", ylab="L1_icache_load_misses",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$dTLB_loads, t$opcode)), type="l", xlab="op_count", ylab="dTLB_loads",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$dTLB_load_misses, t$opcode)), type="l", xlab="op_count", ylab="dTLB_load_misses",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$iTLB_loads, t$opcode)), type="l", xlab="op_count", ylab="iTLB_loads",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+matplot(do.call(data.frame, split(t$iTLB_load_misses, t$opcode)), type="l", xlab="op_count", ylab="iTLB_load_misses",axes=FALSE)
+axis(side=1, at=1:11, labels=c(0,5,10,15,20,25,30,35,40,45,50))
+axis(2)
+```
+
+Looking at the graphs we can say that a few factors have linear dependencies. For the sake of completeness we calculate Pearson correlations.
+
+```{r}
+a <- split(t, t$opcode)
+for (j in 1:length(a)) {
+  n <- names(a)[j]
+  i <- a[j][[n]]
+  b <- cor(i[,-1:-2], i$op_count, method="pearson")
+  colnames(b) <- c(n)
+  print(b)
+}
+```
+
+### task_clock, instructions, L1_dcache_loads, L1_icache_loads, dTLB_loads, iTLB_loads
+
+These stats are strongly correlated. Moreover, the graphs look similar. As an additional observation we check how they are related each other.
+
+For each stat and each opcode we calculate
+
+```
+slope(opcode)/slope(EXP)
+```
+
+where dividing by slope(EXP) is for the normalization purposes. We could think of this as the average cache usage per opcode compared to EXP opcode.
+
+```{r fig.width=10}
+a <- split(t, t$opcode)
+ns <- c("task_clock", "instructions", "L1_dcache_loads", "L1_icache_loads", "dTLB_loads", "iTLB_loads")
+slopes <- list()
+for (j in 1:length(a)) {
+  n <- names(a)[j]
+  i <- a[j][[n]]
+  s <- lapply(paste(ns, '~ op_count'), function(f){
+    b <- lm(as.formula(f), data=i)
+    b$coefficients["op_count"]
+  })
+  if (n == "EXP") {
+    EXPslopes <- s
+  }
+  slopes <- append(slopes, s)
+}
+for (j in 1:length(slopes)) {
+  slopes[[j]] = slopes[[j]] / EXPslopes[[((j-1) %% length(ns)) + 1]]
+}
+slopes <- matrix(slopes, nrow = length(ns), ncol = length(a), byrow = FALSE)
+matplot(slopes, type="l", xlab = "stats", axes=FALSE)
```

Comment:
in this plot, we need a legend to figure out which of the OPCODEs are outliers, because there clearly are some